Essentials/GridFinder.cs

```diff
@@ -0,0 +1,321 @@
+
+                foreach (var group in groups) {
+                    foreach (var node in group.Nodes) {
+
+                        MyCubeGrid grid = node.NodeData;
+
+                        if (grid.Physics == null)
+                            continue;
+
+                        grids.Add(grid);
+                    }
+                }
+
+            } else {
+
+                ConcurrentBag<MyGroups<MyCubeGrid, MyGridMechanicalGroupData>.Group> groups;
+
+                if (gridNameOrEntityId == null)
+                    groups = FindLookAtGridGroupMechanical(character);
+                else
+                    groups = FindGridGroupMechanical(gridNameOrEntityId);
+
+                if (groups.Count > 1)
+                    return null;
+
+                foreach (var group in groups) {
+                    foreach (var node in group.Nodes) {
+
+                        MyCubeGrid grid = node.NodeData;
+
+                        if (grid.Physics == null)
+                            continue;
+
+                        grids.Add(grid);
+                    }
+                }
+            }
+
+            return grids;
+        }
+
+        public static ConcurrentBag<MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Group> FindGridGroup(string gridName) {
+
+            ConcurrentBag<MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Group> groups = new ConcurrentBag<MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Group>();
+            Parallel.ForEach(MyCubeGridGroups.Static.Physical.Groups, group => {
+
+                foreach (MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Node groupNodes in group.Nodes) {
+
+                    MyCubeGrid grid = groupNodes.NodeData;
+
+                    if (grid.Physics == null)
+                        continue;
+
+                    /* Gridname is wrong ignore */
+                    if (!grid.DisplayName.Equals(gridName) && grid.EntityId+"" != gridName)
+                        continue;
+
+                    groups.Add(group);
+                }
+            });
+
+            return groups;
+        }
+
+        public static ConcurrentBag<MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Group> FindLookAtGridGroup(IMyCharacter controlledEntity) {
+
+            const float range = 5000;
+            Matrix worldMatrix;
+            Vector3D startPosition;
+            Vector3D endPosition;
+
+            worldMatrix = controlledEntity.GetHeadMatrix(true, true, false); // dead center of player cross hairs, or the direction the player is looking with ALT.
+            startPosition = worldMatrix.Translation + worldMatrix.Forward * 0.5f;
+            endPosition = worldMatrix.Translation + worldMatrix.Forward * (range + 0.5f);
+
+            var list = new Dictionary<MyGroups<MyCubeGrid, MyGridPhysicalGroupData>.Group, double>();
+            var ray = new RayD(startPosition, worldMatrix.Forward);
+
+            foreach (var group in MyCubeGridGroups.Static.Physical.Groups) {
```

Comment:
See if you can reduce the nesting here by inverting conditionals, it gets pretty deep for my liking